includes/sanitizers/class-amp-story-sanitizer.php

```diff
@@ -20,50 +20,89 @@ class AMP_Story_Sanitizer extends AMP_Base_Sanitizer {
 	 */
 	public static $tag = 'amp-story-page';
 
+	/**
+	 * AMP Story tag spec.
+	 *
+	 * @var array
+	 */
+	private $amp_story_tag_spec;
+
+	/**
+	 * AMP Story Page tag spec.
+	 *
+	 * @var array
+	 */
+	private $amp_story_page_tag_spec;
+
 	/**
 	 * Sanitize the AMP elements contained by <amp-story-page> element where necessary.
 	 *
 	 * @since 0.2
 	 */
 	public function sanitize() {
-		$nodes     = $this->dom->getElementsByTagName( self::$tag );
-		$num_nodes = $nodes->length;
+		$this->amp_story_tag_spec      = AMP_Allowed_Tags_Generated::get_allowed_tag( 'amp-story' )[0];
+		$this->amp_story_page_tag_spec = AMP_Allowed_Tags_Generated::get_allowed_tag( 'amp-story-page' )[0];
 
-		if ( 0 === $num_nodes ) {
-			return;
+		$amp_story_element = $this->dom->getElementsByTagName( 'amp-story' )->item( 0 );
+		if ( $amp_story_element instanceof DOMElement ) {
+			$this->sanitize_story_element( $amp_story_element );
 		}
+	}
 
-		for ( $i = $num_nodes - 1; $i >= 0; $i-- ) {
-			$node = $nodes->item( $i );
+	/**
+	 * Sanitize the children of an amp-story element.
+	 *
+	 * @param DOMElement $element An amp-story element.
+	 */
+	private function sanitize_story_element( DOMElement $element ) {
+		$page_number = 0;
 
-			if ( ! $node ) {
-				continue;
+		$node = $element->firstChild;
+		while ( $node ) {
+			$next_node = $node->nextSibling;
+			if ( $node instanceof DOMElement ) {
```

Comment:
Same logic inversion here, I would prefer an early return (`continue` in this case) instead of making the main logic look like an edge case.

Reply:
True, but the reason why I did it this way was because of `$node = $next_node` needing to run below. Otherwise, I'd have added:

```php
if ( ! $node instanceof DOMElement ) {
    $node = $next_node;
    continue;
}
```

But that seems worse because the logic is duplicated.

Reply:
How about something like this:
```php
$node = $element->firstChild;
do {
	$next_node = $node->nextSibling;
	if ( ! $node instanceof DOMElement ) {
		continue;
	}
	if ( 'amp-story-page' === $node->nodeName ) {
		$page_number++;
		$this->sanitize_story_page_element( $node, $page_number );
	} elseif ( ! in_array( $node->nodeName, $this->amp_story_tag_spec['tag_spec']['child_tags']['child_tag_name_oneof'], true ) ) {
		$this->remove_invalid_child( $node );
	}
} while ( $node = $next_node );
```

Reply:
Note: This is mostly just preference here. I'll approve the changes and let you decide whether you want to make changes or not.

Reply:
Thanks, I like that. However, I tried it and then there is a PHPCS compliant: `WordPress.CodeAnalysis.AssignmentInCondition.FoundInWhileCondition`. We can revisit later.